Bluetooth: Mesh: Add check to LPN friend_req_sent

Andrewpini opened the pull request:
Adds state check to the friend_req_sent CB to ensure
that the CB does not alter the LPN state if it is in a
disabled state. This resolves behavioral issue for lpn_set(false).

Also adds BabbleSim test to check for correct behavior.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

trond-snekvik commented on subsys/bluetooth/mesh/lpn.c:
```diff
@@ -284,6 +284,10 @@ static void friend_req_sent(uint16_t duration, int err, void *user_data)
 		return;
 	}
 
+	if (lpn->state == BT_MESH_LPN_DISABLED) {
```
Could even be `lpn->state != BT_MESH_LPN_ENABLED`, no?

Andrewpini replied:
Yes, if this is the only valid state that the LPN should be in for the CB to complete, it gives more sense. Must confess that I'm not sure if there is any other possible state that is valid, but if you are quite certain then I will change it.

trond-snekvik commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_friendship.c:
```diff
@@ -361,6 +361,20 @@ static void test_friend_group(void)
+ *
+ * Dummy friendship node for test_lpn_disable
+ */
+static void test_friend_lpn_disable_dmy(void)
```
```suggestion
static void test_friend_no_est(void)
```

trond-snekvik commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_friendship.c:
```diff
@@ -361,6 +361,20 @@ static void test_friend_group(void)
+static void test_friend_lpn_disable_dmy(void)
+{
+	/* Wait for LPN node to toggle lpn_set */
+	k_sleep(K_SECONDS(1));
```
No need for this, I believe? We should just be waiting for LPN to send the friend request, which we should be able to receive right away, right?

Andrewpini replied:
I tried this at first, but the simulation environment crashed for the friendship node without the delay for some reason (Before adding the fix to the LPN module):

`p_2G4: WARNING: (src/p2G4_com.c:110): Low level communication with device 1 broken during Abort reevaluation (tried to get 16 got 0 bytes) (most likely the device was terminated)`
`p_2G4: ERROR: (src/p2G4_com.c:112): Exiting`
`timeout: the monitored command dumped core`
`/home/anderss/repos/ncs/zephyr/tests/bluetooth/bsim_bt/bsim_test_mesh/tests_scripts/friendship/../../_mesh_test.sh: line 72:  7128 Segmentation fault      timeout 20 $@`

Hence the workaround. This will of course make the test fail, which is kind of what we want it to do, but the cause will be quite unclear to the user. If you have some notion of how to fix this without the delay I'm all for it, would be much cleaner that way.

Andrewpini replied:
Also uncertain if this behaviour might indicate another potential bug in the friendship module

PavelVPV replied:
Could you check again? It seems to be working fine at my desk with your latest changes and without this `k_sleep`.

trond-snekvik replied:
Might be alleviated by the evt_wait below, perhaps?

PavelVPV replied:
> Might be alleviated by the evt_wait below, perhaps?

Yeah, seems so.

Andrewpini replied:
It works now 👍

Andrewpini replied:
Well I spoke too soon. @PavelVPV could you try to run the test while commenting out the fixes in lpn.c? I belive that i just tested it the second time with the fixes in place, which as described above will not trigger the error.

PavelVPV replied:
Yes, it crashes because lpn tries to process friend offer here after being disabled as the lpn state get overwritten. Backtrace:
> Thread 9 "bs_nrf52_bsim_t" received signal SIGSEGV, Segmentation fault.                                                                           
[Switching to Thread 0xf43fdb40 (LWP 55499)]                           
0x08103e73 in bt_mesh_lpn_friend_offer (rx=0xf43fcf64, buf=0xf43fcf58)   
    at /home/pavel/nordic/ncs/git-repos/zephyrproject/zephyr/subsys/bluetooth/mesh/lpn.c:588
588                     if (!lpn->sub->keys[i].valid) {
(gdb) bt
#0  0x08103e73 in bt_mesh_lpn_friend_offer (rx=0xf43fcf64, buf=0xf43fcf58)
    at /home/pavel/nordic/ncs/git-repos/zephyrproject/zephyr/subsys/bluetooth/mesh/lpn.c:588
#1  0x080c7ae3 in ctl_recv (rx=0xf43fcf64, hdr=4 '\004', buf=0xf43fcf58, seq_auth=0xf43fcee0)
    at /home/pavel/nordic/ncs/git-repos/zephyrproject/zephyr/subsys/bluetooth/mesh/transport.c:923
#2  0x080c8575 in trans_unseg (buf=0xf43fcf58, rx=0xf43fcf64, seq_auth=0xf43fcee0)
    at /home/pavel/nordic/ncs/git-repos/zephyrproject/zephyr/subsys/bluetooth/mesh/transport.c:972
#3  0x080cf13f in bt_mesh_trans_recv (buf=0xf43fcf58, rx=0xf43fcf64)

state:
> (gdb) p lpn->state
$4 = BT_MESH_LPN_WAIT_OFFER

Andrewpini replied:
Ok, seems legit. The question is how we deal with it in the test. After the applied fixes we should hopefully not experience this situation anymore, but that is hardly the point. The point is if it acceptable for the test to fail in this manner if a simmilar  bug re-appears. In other words if its ok that the test fails in a "strange" manner. Im not quite sure, so I leave it to you guys to decide if we want to put in the old fix again, or if it is fine as is.

alxelax replied:
Should assertion be added if it causes a segmentation fault?

Andrewpini replied:
I have discussed this with @PavelVPV, and we have agreed that the test is good as it is.

alxelax replied:
Yep, but I meant lpn.c not test

Andrewpini replied:
> Should assertion be added if it causes a segmentation fault?

I'm not sure really

trond-snekvik commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_friendship.c:
```diff
@@ -361,6 +361,20 @@ static void test_friend_group(void)
+
+	bt_mesh_test_setup();
+	bt_mesh_friend_set(BT_MESH_FEATURE_ENABLED);
+	PASS();
```
Should verify that no friendship is actually established, just like the LPN does.

trond-snekvik commented on tests/bluetooth/bsim_bt/bsim_test_mesh/tests_scripts/friendship/lpn_disable.sh:
```diff
@@ -0,0 +1,10 @@
+
+source $(dirname "${BASH_SOURCE[0]}")/../../_mesh_test.sh
+
+# Check that disabling LPN gives correct behaviour
```
Some additional details would be good, this is testing a very specific scenario :)

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_friendship.c:
```diff
@@ -815,33 +836,51 @@ static void test_other_group(void)
+{
+	bt_mesh_test_setup();
+
+	bt_mesh_lpn_set(true);
+	bt_mesh_lpn_set(false);
```
While I was looking at this test, I found that in the lpn code, the terminated callback can be called even if the established callback hasn't been called yet. This happens when disabling the lpn between friend offer and friend update messages. The terminated callback is called regardless of whether the friendship was established or not (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/lpn.c#L230-L247). This can be reproduced with your test if add some delay (for me that was 400msec) before disabling lpn and commenting out lines 378-380. For me this looks like a bug.

Andrewpini replied:
So what you are saying is that the terminated CB(s) should not trigger unless the LPN node actually has a established connection with a friendship node?

PavelVPV replied:
For me that sounds logically correct.

Andrewpini replied:
Did you get the actual babSim terminated CB to trigger? To me it seems like line https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/lpn.c#L244 prevents the CB from triggering unless there was a friend node connected?

PavelVPV replied:
`lpn->frnd` gets assigned when Friend Offer is received: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/lpn.c#L574. At this point friendship is not established yet (LPN needs to wait for Friend Update: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/lpn.c#L1004).

This is the log when I changed the test as I mentioned in the comment above:

> d_00: @00:00:00.000000  *** Booting Zephyr OS build v2.7.99-3008-gefdb18419891  ***
d_01: @00:00:00.000000  *** Booting Zephyr OS build v2.7.99-3008-gefdb18419891  ***
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 2.7 Build 99
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 2.7 Build 99
d_01: @00:00:00.000000  [00:00:00.000,000] <wrn> bt_id: No static addresses stored in controller
d_00: @00:00:00.000000  [00:00:00.000,000] <wrn> bt_id: No static addresses stored in controller
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: Identity: ED:3B:20:15:18:12 (random)
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: HCI: version 5.3 (0x0c) revision 0x0000, manufacturer 0x0059
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: LMP: version 5.3 (0x0c) subver 0xffff
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> mesh_test: Bluetooth initialized
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: Identity: ED:71:8F:C2:E4:6E (random)
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: HCI: version 5.3 (0x0c) revision 0x0000, manufacturer 0x0059
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: LMP: version 5.3 (0x0c) subver 0xffff
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> mesh_test: Bluetooth initialized
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> mesh_test: Mesh initialized
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_mesh_main: Primary Element: 0x0001
d_00: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_main.bt_mesh_provision: net_idx 0x0000 flags 0x00 iv_index 0x0000
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> mesh_test: Mesh initialized
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_mesh_main: Primary Element: 0x0004
d_01: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_main.bt_mesh_provision: net_idx 0x0000 flags 0x00 iv_index 0x0000
d_01: @00:00:00.302276  [00:00:00.302,246] <inf> test_friendship: LPN: Polling 0x0001 (initial)
d_00: @00:00:00.306702  [00:00:00.306,701] <inf> test_friendship: Friend: Poll from 0x0004
d_00: @00:00:00.306702  [00:00:00.306,701] <inf> test_friendship: Friend: established with 0x0004
d_00: @00:00:00.339845  [00:00:00.339,843] <wrn> bt_mesh_friend: Another request before previous completed sending
d_01: @00:00:00.395508  [00:00:00.395,507] <inf> test_friendship: LPN: terminated with 0x0001
d_00: @00:00:29.300782 INFO: test_friend_no_est PASSED
d_01: @00:00:29.693360 INFO: test_lpn_disable PASSED
d_00: @00:00:29.802735  [00:00:29.802,734] <wrn> bt_mesh_friend: Friendship lost with 0x0004
d_00: @00:00:29.802735  [00:00:29.802,734] <inf> test_friendship: Friend: terminated with 0x0004

There is a message `LPN: terminated with 0x0001`, but there is no preceding message `LPN: established ...`.

I think this could be fixed in a separate PR.

Andrewpini replied:
Ok, I had to tweak my sleep time but i got the same now. It is clearly an issue:

> d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_mesh_main: Primary Element: 0x0001
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_mesh_main: Primary Element: 0x0004
d_01: @00:00:00.003888  [00:00:00.003,875] <dbg> bt_mesh_lpn.lpn_set_state: disabled -> enabled
d_01: @00:00:00.003889  [00:00:00.003,875] <dbg> bt_mesh_lpn.send_friend_req: 
d_01: @00:00:00.004088  [00:00:00.004,058] <dbg> bt_mesh_lpn.friend_req_sent: 
d_01: @00:00:00.004088  [00:00:00.004,058] <dbg> bt_mesh_lpn.lpn_set_state: enabled -> req wait
d_01: @00:00:00.102540  [00:00:00.102,539] <dbg> bt_mesh_lpn.lpn_timeout: state: req wait
d_01: @00:00:00.102540  [00:00:00.102,539] <dbg> bt_mesh_lpn.lpn_set_state: req wait -> wait offer
d_01: @00:00:00.302225  [00:00:00.302,215] <dbg> bt_mesh_lpn.bt_mesh_lpn_friend_offer: recv_win 255 queue_size 16 sub_list_size 3 rssi -35 counter 0
d_01: @00:00:00.302225  [00:00:00.302,215] <dbg> bt_mesh_lpn.send_friend_poll: lpn->sent_req 0x00
d_01: @00:00:00.302735  [00:00:00.302,734] <dbg> bt_mesh_lpn.clear_friendship: force 0 disable 1, current state: wait offer
d_01: @00:00:00.302735  [00:00:00.302,734] <dbg> bt_mesh_lpn.lpn_set_state: wait offer -> disabled
d_01: @00:00:00.302735  [00:00:00.302,734] <inf> test_friendship: LPN: terminated with 0x0001
d_00: @00:00:00.306702  [00:00:00.306,701] <inf> test_friendship: Friend: Poll from 0x0004
d_00: @00:00:00.306702  [00:00:00.306,701] <inf> test_friendship: Friend: established with 0x0004
d_00: @00:00:00.339845  [00:00:00.339,843] <wrn> bt_mesh_friend: Another request before previous completed sending
d_00: @00:00:29.802735  [00:00:29.802,734] <wrn> bt_mesh_friend: Friendship lost with 0x0004
d_00: @00:00:29.802735  [00:00:29.802,734] <inf> test_friendship: Friend: terminated with 0x0004

The LPN is disabled while it is waiting for an offer, even though it already has received one from the friend node...

Yes, I agree that this should be a separate PR. Do you have a clear idea how to fix it, or should i give it a go?

PavelVPV replied:
> Do you have a clear idea how to fix it, or should i give it a go?

Not at the moment. You could give it a go if you have a time, or it can be investigated later.

trond-snekvik replied:
That's a good catch, I agree that this shouldn't happen. Can be a separate PR though

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_friendship.c:
```diff
@@ -815,33 +836,51 @@ static void test_other_group(void)
+	bt_mesh_lpn_set(true);
+	bt_mesh_lpn_set(false);
+
+	if (!evt_wait(LPN_ESTABLISHED, K_SECONDS(30))) {
```
By looking at the test description in `lpn_disable.sh` and the lpn code, I think, the test could be more strict if it checks that no `LPN_POLLED` event fired. When establishing a friendship, Friend poll message is sent after Friend Offer message is received and the poll callback is called in this case. Checking `LPN_POLL` event will guarantee the _immediate_ reaction on `bt_mesh_lpn_set(false);` call. This still may not check the introduced fix if the underlying layers' behavior gets changed  and `friend_req_sent` callback is called right in the `bt_mesh_ctl_send` call.

PavelVPV commented on subsys/bluetooth/mesh/lpn.c:
```diff
@@ -284,6 +284,10 @@ static void friend_req_sent(uint16_t duration, int err, void *user_data)
 		return;
 	}
 
+	if (lpn->state != BT_MESH_LPN_ENABLED) {
```
I guess the same check needs to be added into `req_sent` callback.

PavelVPV replied:
Just a small nitpicking: maybe put this check before the log message above?

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_friendship.c:
```diff
@@ -361,6 +361,27 @@ static void test_friend_group(void)
+
+	bt_mesh_test_setup();
+	bt_mesh_friend_set(BT_MESH_FEATURE_ENABLED);
+
```
Extra empty line

Andrewpini commented:
> @Andrewpini It is not clear how it is guaranteed that the friend node is called first. If LPN is called first then the test will pass even functionality doesn't work, does it?

Well the central aspect is that we want to check that the LPN node does not incorrectly overwrites its own state. If the interface allowed it we would not be dependent on a secondary device in this test. Then we could rather check that the LPN node is in a disabled state at the end of the test. The purpose of the Friend node is that if the LPN node fails to enter a disabled state it will eventually initiate a connection to the friendship node, thus failing the test.

ludvigsj reviewed (approved):
Apart from comments raised by Pavel and Trond, this looks okay to me

trond-snekvik reviewed (approved):
Looks good, thanks for taking the time to make a fix! 

I think you've picked up some build issue from main, a rebase should probably resolve it.

Andrewpini committed:
Bluetooth: Mesh: Add LPN-disable BabbleSim test

Adds test to check correct behaviour for disabling
LPN feature.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
Bluetooth: Mesh: Add check to LPN friend_req_sent

Adds state check to the friend_req_sent CB to ensure
that the CB does not alter the LPN state if it is in a
disabled state. This resolves behavioural issue for lpn_set(false).

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>
